### PR TITLE
[Sanchari] Duke Increments

### PR DESCRIPTION
**Pull request by sancharib-git**

### CODE REVIEW

File: src/main/java/Deadline.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package main.java;
+
+public class Deadline extends Task {
+    protected String by;
```
Comment (ElysiaTanZY) [inline src/main/java/Deadline.java:4]: Might want to consider changing the variable name to deadline because it wasn't very clear what this variable was referring to at first glance.

File: src/main/java/Duke.java [hunk 2]
```diff
@@ -1,10 +1,186 @@
...
 public class Duke {
-    public static void main(String[] args) {
-        String logo = " ____        _        \n"
-                + "|  _ \\ _   _| | _____ \n"
-                + "| | | | | | | |/ / _ \\\n"
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+
+    /**Declaration of variables */
```
Comment (ElysiaTanZY) [inline src/main/java/Duke.java]: The comments should be indented relative to their position in the code.

File: src/main/java/Duke.java [hunk 3]
```diff
@@ -1,10 +1,186 @@
...
+        Scanner sc = new Scanner(System.in);
+        ArrayList<Task> Tasks = new ArrayList<>();
+        final String FILEPATH;
+        final File FILE;
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
+        Storage storage;
+
+    /**
+     * Constructor
```
Comment (ElysiaTanZY) [inline src/main/java/Duke.java:22]: There should be an empty line between the method description and parameter section.

File: src/main/java/Duke.java [hunk 4]
```diff
@@ -1,10 +1,186 @@
...
+     * @param String filepath
+     */
+    public Duke(String filepath) {
+        this.FILEPATH = filepath;
+        this.FILE = new File(filepath);
+        this.storage = new Storage(filepath);
     }
+
+    /**
+     * Main method
```
Comment (ElysiaTanZY) [inline src/main/java/Duke.java]: There should be an empty line between the method description and parameter section.

File: src/main/java/Duke.java [hunk 5]
```diff
@@ -1,10 +1,186 @@
...
+    public static void main(String[] args) throws DukeException {
+        
+        Duke d = new Duke("data/list.txt");
+
+        d.run();
+    }
+
+    /**
+     * Run program
+     * @throws DukeException
```
Comment (ElysiaTanZY) [inline src/main/java/Duke.java]: Perhaps you might want to add a bit more explanation about the DukeException, such as when it will be thrown etc.

File: src/main/java/Duke.java [hunk 6]
```diff
@@ -1,10 +1,186 @@
...
+     */
+    public void run() throws DukeException {
+        /** Welcome Message */
+        DukeGreeting dg = new DukeGreeting();
+        dg.showWelcomeMessage();
+        dg.showInstructions();
```
Comment (ElysiaTanZY) [inline src/main/java/Duke.java]: I suggest having a blank line after this line of code because logical blocks should be distinctly separated from each other to enhance readability of your code.

File: src/main/java/Duke.java [hunk 7]
```diff
@@ -1,10 +1,186 @@
...
+        while(true) {
```
Comment (ElysiaTanZY) [inline src/main/java/Duke.java]: According to the Java Coding Standard given to us, this should be while (true) { instead.

File: src/main/java/Duke.java [hunk 8]
```diff
@@ -1,10 +1,186 @@
...
+            String line = sc.nextLine();
+            Ui ui = new Ui(line);
+            String command = ui.getCommand();
+            if (command.equals("bye")) {
```
Comment (ElysiaTanZY) [inline src/main/java/Duke.java]: I would suggest having an empty line before this line to enhance the readability of your code.

File: src/main/java/Duke.java [hunk 9]
```diff
@@ -1,10 +1,186 @@
...
+     * @throws DukeException
+     */
+    public void run() throws DukeException {
+        /** Welcome Message */
+        DukeGreeting dg = new DukeGreeting();
+        dg.showWelcomeMessage();
+        dg.showInstructions();
+        while(true) {
+            String line = sc.nextLine();
+            Ui ui = new Ui(line);
```
Comment (ElysiaTanZY) [inline src/main/java/Duke.java]: Instead of having the Ui class handle both input/output interactions and making sense of user command, I would suggest having a parser class to take care of the latter because it would be clearer if each class just takes care of one thing.

File: src/main/java/Ui.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    public void showList() {
+        System.out.println("Here are the tasks in your list: ");
+    }
+
+    public String errmessage() {
+        return "Invalid Input, Please check instructions";
+    }
+
+    public void doneTask(Task task) {
+        System.out.println("-------------------------------------------------------------\nNice I have marked this task as done\n" +
```
Comment (ElysiaTanZY) [inline src/main/java/Ui.java]: This line is pretty long, please keep each line to a maximum of 120 characters.

File: src/main/java/Storage.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * list tasks from file
+     * @param filepath
+     * @throws FileNotFoundException
+     */
+    public void printFileContents(String filepath) throws FileNotFoundException {
+        File f = new File(filepath); // create a File for the given file path
+        Scanner s = new Scanner(f); // create a Scanner using the File as the source
+        while (s.hasNext()) {
```
Comment (ElysiaTanZY) [inline src/main/java/Storage.java]: I would suggest having an empty line above to enhance the readability of your code.

File: src/main/java/Storage.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        fw.write(textToAdd);
+        fw.close(); 
+    }
+
+    /**
+     * list tasks from file
+     * @param filepath
+     * @throws FileNotFoundException
+     */
+    public void printFileContents(String filepath) throws FileNotFoundException {
```
Comment (ElysiaTanZY) [inline src/main/java/Storage.java]: I would suggest having this method within the Ui class since the Ui class is the one that should be handling input/output interactions.

File: src/main/java/Ui.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
+package main.java;
+
+import java.util.Scanner;
+
+public class Ui {
+    String line;
```
Comment (ElysiaTanZY) [inline src/main/java/Ui.java:4]: Attributes should be declared private.

File: src/main/java/DukeGreeting.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+        System.out.println();
+    }
+
+    /**Methods to show Instructions on user inputs */
+    public void showInstructions() {
+        System.out.println("-------------------------------------------------------------");
+        System.out.println("Instructions: Type in Commands");
+        System.out.println("Examples:");
+        System.out.println("1. ~todo task~" + '\n' + "2. ~deadline task /by yyyy-MM-dd HH:mm~" + '\n' 
+        + "3. ~event task /at yyyy-MM-dd HH:mm~" + '\n' + "4. ~delete taskname~" + '\n' + "5. ~done taskname~" + '\n' + "6. ~list~");
```
Comment (ElysiaTanZY) [inline src/main/java/DukeGreeting.java]: This line is pretty, please try to keep to a maximum of 120 characters per line.

File: src/main/java/Duke.java [hunk 15]
```diff
@@ -1,10 +1,186 @@
...
+                    LocalDateTime ldt = LocalDateTime.parse(date, formatter);
+                    String sldt = ldt.format(formatter);
+                    System.out.println("Here are the tasks in your list for this date:");
+                    System.out.println(storage.getTask(sldt, FILEPATH));
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+         } else if(command.equals("find")) {
+            try {
+                String rest = ui.getDescription();
```
Comment (ElysiaTanZY) [inline src/main/java/Duke.java]: Might want to change the variable name to 'keyword' instead because it is not very clear what this variable is referring to at first glance.

File: src/main/java/Storage.java [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    /**
+     * retrieve task from storage
+     * @param lineContent
+     * @param filepath
+     * @return String task
+     * @throws IOException
+     */
+    public String getTask(String lineContent, String filepath) throws IOException {
```
Comment (ElysiaTanZY) [inline src/main/java/Storage.java]: Since this class already stores the file path as one of its attributes and the same file path is always used, perhaps you might want to consider removing the parameter 'file path' from the methods in this class.

File: src/main/java/Duke.java [hunk 17]
```diff
@@ -1,10 +1,186 @@
...
+                }
+            } else if (command.equals("done")) {
+                if (ui.getDescription().equals(null)) {
+                    throw new DukeException(ui.errmessage());
+                } else {
+                    String desc = ui.getDescription();
+                    Task t = new Task(desc);
+                    t.markAsDone();
+                    try {
+                        String upd8task = storage.getTask(desc, FILEPATH).replaceAll("["+"\u2718"+"]", "["+"\u2713"+"]");
```
Comment (ElysiaTanZY) [inline src/main/java/Duke.java]: From what I understand, having this line means that I cannot have two completely different tasks but with the same description. E.g. If I have 2 tasks (A and B) with description 'CS2103 Assignment' then marking task A as done with mark task B as done as well. Perhaps, you might want to find a different method to identify tasks that are to be marked as done (E.g. By index number as suggested in the project description)

**Review by ElysiaTanZY — COMMENTED** [review]
A few additional comments that pertain to your code in general: 
1. There should be punctuations behind each parameter description. 
2. There should be an empty line to separate the method description and parameter section. 
3. Class and their public methods require header comments, parameters should be given some brief descriptions as well. 
4. Might want to consider loading the tasks into an array list at the start for easier manipulation rather than having to access the file for every single command. Implement the TaskList class as well. 
5. Consider implementing the Parser class to make sense of the user input rather than getting the Ui class to do everything. 
6. Consider adding empty lines between logical blocks of code and adhere to the white space requirements of the Coding Standard given to us to make your code more readable.

File: src/main/java/DukeGreeting.java [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+    /**Method to show goodbye message */
+    public void showGoodbyeMessage() {
+        System.out.println("-------------------------------------------------------------");
+                System.out.println("Bye. Hope to see you again soon!");
+                System.out.println();
+                System.out.println("-------------------------------------------------------------");
+                System.out.println();
+    }
+}
```
Comment (ongch00) [inline src/main/java/DukeGreeting.java]: Class and method javadocs should be the first form and not the single line form.

File: src/main/java/Task.java [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+     * @return String description
+     */
+    public String getDescription() {
+        return this.description;
+    }
+
+    /**
+     * getter for status icon
+     * @return String status icon
+     */
```
Comment (ongch00) [inline src/main/java/Task.java]: for the javadoc, the summary should be something like Gets [something]

File: src/main/java/Event.java [hunk 20]
```diff
@@ -0,0 +1,24 @@
+package main.java;
+
+public class Event extends Task {
+    protected String at;
```
Comment (ongch00) [inline src/main/java/Event.java:4]: variable name could be more descriptive.

File: src/main/java/Event.java [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+
+    public Event(String description, String at) {
+        super(description);
+        this.at = at;
+    }   
+
+
+    /**
+     * method to get Date from this.task
+     * @return String date
```
Comment (ongch00) [inline src/main/java/Event.java]: There should be an empty line between the description and the return

File: src/main/java/Storage.java [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+import java.io.File;
+import java.nio.file.FileAlreadyExistsException;
+import java.io.FileNotFoundException;
+import java.util.stream.*;
+import java.util.ArrayList;
+import java.util.Scanner;
+
+public class Storage{
+
+    String FILEPATH;
```
Comment (ongch00) [inline src/main/java/Storage.java:8]: this should follow the camelCase naming style

**Review by ongch00 — COMMENTED** [review]
Some Javadocs and code styling issues to take note of. Other than that, keep up the good work!

File: src/main/java/Ui.java [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+
+    public String showList() {
+        return ("Here are the tasks in your list: ");
+    }
+
+
+
+    public String errmessage() {
```
Comment (duongphammmm) [inline src/main/java/Ui.java]: Method name should be `showErrorMessage`

File: src/main/java/Duke.java [hunk 24]
```diff
@@ -1,10 +1,329 @@
...
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.FileAlreadyExistsException;
+import java.io.FileNotFoundException;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.LocalDate;
+import java.time.temporal.ChronoUnit;
+import java.time.LocalDateTime;
+import java.util.stream.*;
```
Comment (duongphammmm) [inline src/main/java/Duke.java]: Try to avoid wildcard imports using `*` and instead only import classes that you need.

File: src/main/java/Duke.java [hunk 25]
```diff
@@ -1,10 +1,329 @@
...
+        mainLayout.setPrefSize(400.0, 600.0);
+
+        scrollPane.setPrefSize(385, 535);
+        scrollPane.setHbarPolicy(ScrollPane.ScrollBarPolicy.NEVER);
+        scrollPane.setVbarPolicy(ScrollPane.ScrollBarPolicy.ALWAYS);
+
+        scrollPane.setVvalue(1.0);
+        scrollPane.setFitToWidth(true);
+
+        // You will need to import `javafx.scene.layout.Region` for this.
```
Comment (duongphammmm) [inline src/main/java/Duke.java:75]: Maybe consider to remove comments given in the javafx guide like this one? :)

File: src/main/java/Duke.java [hunk 26]
```diff
@@ -1,10 +1,329 @@
...
+                tl.TL.add(todo);
+                response = ui.addedTask(todo);
+            } catch (Exception e) {
+                response = e.getMessage();
+            }
+        } else if (command.equals("deadline")) {
+            try {
+                //response = "deadline";
+                assert isTaskListLoaded : "TaskList not loaded from storage";
+                String desc = ui.getDescription();
```
Comment (duongphammmm) [inline src/main/java/Duke.java]: Quite a few number of if-else clauses that make it hard to follow. Consider using some form of abstraction (SLAP) or use switch statements to improve readability!

File: src/main/java/DukeException.java [hunk 27]
```diff
@@ -0,0 +1,17 @@
+package main.java;
+
+/**
+ * Class to handle exceptions with Duke program
+ */
+public class DukeException extends Exception {
+
+    private static final long serialVersionUID = 1L;
```
Comment (duongphammmm) [inline src/main/java/DukeException.java:8]: Constant variables are conventionally represented by uppercase name, separated by underscore (e.g.: SERIAL_VERSION_UID)

File: src/main/java/Storage.java [hunk 28]
```diff
@@ -0,0 +1,123 @@
...
+     * @param lineContent
+     * @param filepath
+     * @return String task
+     * @throws IOException
+     */
+    public String getTask(String lineContent, String filepath) throws IOException {
+        File file = new File(filepath);
+        return Files.lines(file.toPath())
+                    .filter(line -> line.contains(lineContent))
+                    .reduce("", (x,y) -> x+y+'\n');
```
Comment (duongphammmm) [inline src/main/java/Storage.java]: There should be white spaces around operators like `+`

File: src/main/java/Storage.java [hunk 29]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    /**
+     * list tasks from file
+     * @param filepath
+     * @throws FileNotFoundException
+     */
+    public void printFileContents(String filepath) throws FileNotFoundException {
+        File f = new File(filepath); // create a File for the given file path
+        Scanner s = new Scanner(f); // create a Scanner using the File as the source
```
Comment (duongphammmm) [inline src/main/java/Storage.java]: Personally i would prefer naming my variable using a longer version, e.g. `scanner` instead of `s`. The variable is used quite shortly after this so it is still okay, but if after about 20 lines of code, i would probably forget what `s` stands for :)

File: src/main/java/Task.java [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Constructor
+     * @param description
+     */
+    public Task(String description) {
+        if(description.contains("["+"\u2713"+"]") || description.contains("["+"\u2718"+"]"))
+         this.description = description;
+        else 
+        this.description = "["+"\u2718"+"]" + description;
```
Comment (duongphammmm) [inline src/main/java/Task.java:15]: Remember to have your brackets :)

**Review by duongphammmm — COMMENTED** [review]
Overall, I really like how your codes are generally clean and readable with good logic. Just need to be mindful of small code quality violations :D

File: src/main/java/Duke.java (hunk 5 above)
Comment (hckim0911) [reply src/main/java/Duke.java]: Yes, for details on your javadoc, you can think about adding more comments here :)

File: src/main/java/Ui.java (hunk 13 above)
Comment (hckim0911) [reply src/main/java/Ui.java:4]: Someone sometimes follows this style in the industry because of its default attribute on this. Maybe you can check about it later :)

File: src/main/java/DukeGreeting.java (hunk 14 above)
Comment (hckim0911) [reply src/main/java/DukeGreeting.java]: Agree, you can add "carriage return" to increase readability on your codes.

File: src/main/java/Duke.java (hunk 26 above)
Comment (hckim0911) [reply src/main/java/Duke.java]: I agree with this comment.

**Comment by hckim0911** [comment]
I have gone through your codes and saw comments made by other students. I think they made good comments which are useful for increasing the quality of your codes. :) You can reflect them on your codes as much as you can.
